tough-kms/src/lib.rs

```diff
@@ -188,7 +196,130 @@ impl Sign for KmsRsaKey {
                 profile: self.profile.clone(),
                 key_id: self.key_id.clone(),
             })?;
-        let signature = response.signature.context(error::SignatureNotFound)?;
-        Ok(signature.to_vec())
+        let signature = response
+            .signature
+            .context(error::SignatureNotFound)?
+            .to_vec();
+
+        // sometimes KMS produces a signature that is shorter than the modulus. in those cases,
+        // we have observed that openssl and KMS will both validate the signature, but ring will
+        // not. if we pad the beginning of the signature with zeros to make the signature exactly
+        // the same length as the modulus, then ring will verify the signature.
+        let signature = pad_signature(signature, self.modulus_size_bytes)?;
+        Ok(signature)
+    }
+}
+
+/// Parses the `CustomerMasterKeySpec` string returned by KMS, e.g. `RSA_3072` and returns the size
+/// of the modulus in bytes. For example `RSA_3072` has a modulus of 3072 bits, so the function will
+/// return 384 == (3072 / 8). If the parsed number is not divisible by 8, an error is returned.
+fn parse_modulus_length_bytes(spec: &str) -> error::Result<usize> {
+    // only RSA is currently supported
+    ensure!(
+        spec.starts_with("RSA_"),
+        error::BadCustomerMasterKeySpec { spec }
+    );
+    // prevent a panic if the string is precisely "RSA_"
+    ensure!(spec.len() > 4, error::BadCustomerMasterKeySpec { spec });
+    // extract the digits
+    let mod_len_str = &spec[4..];
+    // parse the digits
+    let mod_bits = mod_len_str
+        .parse::<usize>()
+        .context(error::BadCustomerMasterKeySpecInt { spec })?;
+    // make sure the modulus size is compatible with u8 bytes
```

Comment:
Do you think instead of parsing `customer_master_key_spec` we can use enum that maps `customer_master_key_spec` to number of bytes. That way we won't need to parse string and can avoid unsupported customer_master_key_spec from being used.

Reply:
I think I would prefer to be agnostic about the bitsize, provided it is divisible by 8, rather than maintaining a variant. That way if KMS adds support for a size, it will still 'just work', instead of falsely reporting that the bit size is unsupported.

Reply:
Another thought can we use signing algorithm: `RSASSA_PSS_SHA_256 | RSASSA_PSS_SHA_384 | RSASSA_PSS_SHA_512 | RSASSA_PKCS1_V1_5_SHA_256 | RSASSA_PKCS1_V1_5_SHA_384 | RSASSA_PKCS1_V1_5_SHA_512 | ECDSA_SHA_256 | ECDSA_SHA_384 | ECDSA_SHA_512`
to extract bit size because it has consistent naming for different algo, instead of `CustomerMasterKeySpec` which has `RSA_2048 | RSA_3072 | RSA_4096 | ECC_NIST_P256 | ECC_NIST_P384 | ECC_NIST_P521 | ECC_SECG_P256K1 | SYMMETRIC_DEFAULT` where bit size naming is not consistent. Either way feels bad to me :(, because we are dependent on a naming convention. I searched if there is a way to identify kms key bit size, but couldn't find anything.

Reply:
The algorithm is independent of the size of the key modulus. e.g. RSASSA_PSS_SHA_256 works with 2048, 3072, etc. `CustomerMasterKeySpec` is the only place I know of that we can get the key size. I'm not sure how the ring algorithm figures it out, I guess it must be encoded in the x509_spki representation of the public key, but that seems like a more difficult place to parse it from.